website_sale_product_contract_gift/tests/test_gift.py

```diff
@@ -0,0 +1,102 @@
+# SPDX-FileCopyrightText: 2024 Coop IT Easy SC
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+
+from odoo import Command, fields
+from odoo.tests import common
+
+
+class TestGiftContractBase(common.TransactionCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.today = fields.Date.today()
+        cls.partner = cls.env["res.partner"].create(
+            {
+                "name": "partner test contract",
+                "email": "demo@demo.com",
+            }
+        )
+        cls.partner_gift_to = cls.env["res.partner"].create(
+            {
+                "name": "partner gift to",
+                "email": "partner_gift_to@demo.com",
+                "type": "delivery",
+            }
+        )
+        cls.product_1 = cls.env.ref("product.product_product_1")
+        cls.product_1.is_gift = True
+        cls.product_2 = cls.env.ref("product.product_product_2")
+        cls.product_2.is_gift = False
+
+
+class TestGiftContract(TestGiftContractBase):
+    def _get_gift_sale_order(self):
+        order = self.env["sale.order"].create(
+            {
+                "partner_id": self.partner.id,
+                "partner_shipping_id": self.partner_gift_to.id,
+                "order_line": [
+                    Command.create(
+                        {
+                            "product_id": self.product_1.id,
+                            "date_start": self.today,
+                        }
+                    ),
+                ],
+            }
+        )
+        return order
+
+    def _generate_contract_from_order(self):
+        order = self._get_gift_sale_order()
+        order.action_confirm()
+        contract = order.order_line.mapped("gift_contract_id")[0]
```

Comment:
why not simply:
```suggestion
        contract = order.order_line[0].gift_contract_id
```

Reply:
For me these two line are not totally equivalent.

Reply:
`mapped` will filter line that have a `gift_contract_id`. But can be empty and selecting line 0 can fail in such a case.

Reply:
my suggestion is to only used `.mapped()` when needed (that is: to access non-relational fields on a recordset of potentially multiple records).

i’m not sure i understand your last explanation. can you explain a case where the code with `.mapped()` succeeds but not the one without?

Reply:
With my code I get the first contract on an order line that have a contract. Because "mapped" will return all the contract found on the order line. I will not return empty recordset for the line that don’t have a contract.

In your code, you take the first order line (and you assume there is a contract on it) and you take the contract from this line.

Maybe a equivalent without `mapped` can be:
```suggestion
        contract = order.order_line.gift_contract_id[0]
```

Reply:
done.